pkg/personname/groupInfo.go

```diff
@@ -0,0 +1,170 @@
+	"strings"
+)
+
+const segmentSep = "^"
+
+// GroupTrailingNullLevel represents how many null '^' separators are present in the
+// GroupInfo.DCM() return value.
+type GroupTrailingNullLevel uint
+
+// String implements fmt.Stringer, giving human-readable names to the trailing null
+// level.
+//
+// Returns "NONE" if no null separators were present.
+//
+// Returns "ALL" if the highest possible null separator was present.
+//
+// Otherwise, returns the name of the section that comes after the highest present null
+// separator.
+//
+// String will panic if called on a value that exceeds GroupNullAll.
+func (level GroupTrailingNullLevel) String() string {
+	switch level {
+	case GroupNullNone:
+		return "NONE"
+	case GroupNullGiven:
+		return "GivenName"
+	case GroupNullMiddle:
+		return "MiddleName"
+	case GroupNullPrefix:
+		return "NamePrefix"
+	case GroupNullAll:
+		return "ALL"
+	default:
+		panic(validateGroupNullSepLevel(level))
+	}
+}
+
+const (
+	// GroupNullNone will render no null seps.
+	GroupNullNone GroupTrailingNullLevel = iota
+
+	// NullSepGiven will render null separators up to the separator before the
+	// GroupInfo.GivenName segment
+	GroupNullGiven
+
+	// NullSepGiven will render null separators up to the separator before the
+	// GroupInfo.MiddleName segment
+	GroupNullMiddle
+
+	// NullSepGiven will render null separators up to the separator before the
+	// GroupInfo.NamePrefix segment
+	GroupNullPrefix
+
+	// NullSepGiven will render null separators up to the separator before the
+	// GroupInfo.NameSuffix segment (ALL possible separators).
+	GroupNullAll
+)
+
+func validateGroupNullSepLevel(level GroupTrailingNullLevel) error {
+	if level <= GroupNullAll {
+		return nil
+	}
+
+	return newErrNullSepLevelInvalid(uint(GroupNullAll), uint(level))
+}
+
+// GroupInfo holds the parsed information for any one of these groups the person name
+// groups specified in the DICOM spec:
+//
+//	- Alphabetic
+//	- Ideographic
+//	- Phonetic
+type GroupInfo struct {
+	// FamilyName is the person's family or last name.
+	FamilyName string
+	// GivenName if the person's given or first names.
+	GivenName string
+	// MiddleName is the person's middle name(s).
+	MiddleName string
+	// NamePrefix is the person's name prefix (ex: MR., MRS.).
+	NamePrefix string
+	// NameSuffix is the person's name suffix (ex: Jr, III).
+	NameSuffix string
+
+	// TrailingNullLevel contains the highest present null '^' separator in the DCM()
+	// value. For most use cases GroupNullAll or GroupNullNone should be used when
+	// creating new PN values. Use other levels only if you know what you are doing!
+	TrailingNullLevel GroupTrailingNullLevel
+}
+
+// DCM Returns original, formatted string in
+// '[FamilyName]^[GivenName]^[MiddleName]^[NamePrefix]^[NameSuffix]'.
+func (group GroupInfo) DCM() string {
+	// validate our TrailingNullLevel and panic if it is exceeded.
+	if err := validateGroupNullSepLevel(group.TrailingNullLevel); err != nil {
+		panic(err)
```

Comment:
I'm not so sure about panicing here. Either we should return an error (which may be slighly annoying to add to this API) or revert to a sane default.

See https://github.com/golang/go/wiki/CodeReviewComments#dont-panic

Perhaps returning an error is the best way to go. WDYT?

Reply:
I agree that in general panics should be avoided. My feeling here was that having an invalid null value is an Undefined behavior: like reading a slice index that does not exist.

And, as you intuited, my thought was that adding an error return to this method makes it a little unwieldy.

I think let's add an error, as well as a MustDCM() method for when you know you won't have to deal with them.